### PR TITLE
Feature/av 4070 avetmiss 8 support

### PR DESCRIPTION
**Pull request by SyntheticDave**
Adds support for processing v8 avetmiss data files.

Each new store is added in its own commit. Specs for all new stores are tidied in 4cf9880.

### CODE REVIEW

File: lib/avetmiss_data.rb [hunk 1]
```diff
@@ -18,6 +21,9 @@ module V7
 require 'avetmiss_data/package'
 require 'avetmiss_data/parser'
 require 'avetmiss_data/builder'
+require 'avetmiss_data/stores/concerns/v7_v8_rto_delivery_location'
```
Comment (schlick) [inline lib/avetmiss_data.rb]: Do not share code between V7 and V8 as it is coincidental duplicate.

File: lib/avetmiss_data.rb [hunk 2]
```diff
@@ -18,6 +21,9 @@ module V7
...
+require 'avetmiss_data/stores/concerns/v7_v8_disability'
```
Comment (schlick) [inline lib/avetmiss_data.rb]: Ditto. Split up into separate V7 and V8.

File: lib/avetmiss_data.rb [hunk 3]
```diff
@@ -18,6 +21,9 @@ module V7
...
+require 'avetmiss_data/stores/concerns/v7_v8_achievement'
```
Comment (schlick) [inline lib/avetmiss_data.rb]: Ditto. Split up into separate V7 and V8.

File: lib/avetmiss_data/package.rb [hunk 4]
```diff
@@ -13,6 +12,14 @@ class AvetmissData::Package
...
   }
 
+  V7_FILES_MAP = {
+    course_stores: "NAT00030",
+  }.merge(FILES_MAP)
+
+  V8_FILES_MAP = {
+    course_stores: "NAT00030A",
+  }.merge(FILES_MAP)
+
```
Comment (schlick) [inline lib/avetmiss_data/package.rb]: It's probably irrelevant but V6 is being ignored here.

File: lib/avetmiss_data/package.rb [hunk 5]
```diff
@@ -30,9 +37,11 @@ class AvetmissData::Package
   attr_accessor :achievement_stores
   attr_accessor :enrolment_stores
   attr_accessor :qual_completion_stores
+  attr_accessor :version
 
-  def initialize(attributes = {})
+  def initialize(attributes = {}, version: :v7)
```
Comment (schlick) [inline lib/avetmiss_data/package.rb]: Two options:

* default to V8, or
* have no default and always require the caller to supply the version

Either of these changes would be a breaking change so bumping to v2.0.0 would be recommended.

File: lib/avetmiss_data/package.rb [hunk 6]
```diff
@@ -127,7 +136,8 @@ def self.from_zip_file(zip_file, attributes={})
   end
 
   def to_zip_file
-    generate_zip_file(FILES_MAP)
+    files_map = self.version == :v7 ? V7_FILES_MAP : V8_FILES_MAP
```
Comment (schlick) [inline lib/avetmiss_data/package.rb]: We're not taking into account V6. We probably don't need to but I'd at least prefer the default to be V8.

File: lib/avetmiss_data/stores/concerns/v7_v8_achievement.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  included do
+    nat_file('NAT00100', {
+      client_identifier: 0...10,
+      prior_education_achievement_identifier: 10...13,
+      extras: 13..-1
+    })
+
+    alias_method :identifier, :client_identifier
+  end
+end
```
Comment (schlick) [inline lib/avetmiss_data/stores/concerns/v7_v8_achievement.rb]: Split.

File: lib/avetmiss_data/stores/concerns/v7_v8_disability.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  included do
+    nat_file('NAT00090', {
+      client_identifier: 0...10,
+      disability_type_identifier: 10...12,
+      extras: 12..-1
+      })
+
+      alias_method :identifier, :client_identifier
+  end
+end
```
Comment (schlick) [inline lib/avetmiss_data/stores/concerns/v7_v8_disability.rb]: Split

File: lib/avetmiss_data/stores/concerns/v7_v8_rto_delivery_location.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      post_code: 120...124,
+      state_identifier: 124...126,
+      address_location: 126...176,
+      country_identifier: 176...180,
+      extras: 180..-1
+      })
+
+      alias_method :identifier, :training_organisation_identifier
+  end
+end
```
Comment (schlick) [inline lib/avetmiss_data/stores/concerns/v7_v8_rto_delivery_location.rb]: Split

File: lib/avetmiss_data/stores/v7/achievement.rb [hunk 10]
```diff
@@ -1,9 +1,3 @@
 class AvetmissData::Stores::V7::Achievement < AvetmissData::Stores::V7::Base
-  nat_file('NAT00100', {
-    client_identifier: 0...10,
-    prior_education_achievement_identifier: 10...13,
-    extras: 13..-1
-  })
-
-  alias_method :identifier, :client_identifier
```
Comment (schlick) [inline lib/avetmiss_data/stores/v7/achievement.rb]: Reinstate.

File: lib/avetmiss_data/stores/v7/disability.rb [hunk 11]
```diff
@@ -1,9 +1,3 @@
 class AvetmissData::Stores::V7::Disability < AvetmissData::Stores::V7::Base
-  nat_file('NAT00090', {
-    client_identifier: 0...10,
-    disability_type_identifier: 10...12,
-    extras: 12..-1
-  })
-
-  alias_method :identifier, :client_identifier
```
Comment (schlick) [inline lib/avetmiss_data/stores/v7/disability.rb]: Reinstate.

File: lib/avetmiss_data/stores/v7/rto_delivery_location.rb [hunk 12]
```diff
@@ -1,14 +1,3 @@
...
-    training_organisation_delivery_location_identifier: 10...20,
-    training_organisation_delivery_location_name: 20...120,
-    post_code: 120...124,
-    state_identifier: 124...126,
-    address_location: 126...176,
-    country_identifier: 176...180,
-    extras: 180..-1
-  })
-
-  alias_method :identifier, :training_organisation_identifier
```
Comment (schlick) [inline lib/avetmiss_data/stores/v7/rto_delivery_location.rb]: Reinstate.

File: lib/avetmiss_data/stores/v8/achievement.rb [hunk 13]
```diff
@@ -0,0 +1,3 @@
+class AvetmissData::Stores::V8::Achievement < AvetmissData::Stores::V8::Base
+  include AvetmissData::Stores::V7V8Achievement
```
Comment (schlick) [inline lib/avetmiss_data/stores/v8/achievement.rb]: Extract back out.

File: lib/avetmiss_data/stores/v8/disability.rb [hunk 14]
```diff
@@ -0,0 +1,3 @@
+class AvetmissData::Stores::V8::Disability < AvetmissData::Stores::V8::Base
+  include AvetmissData::Stores::V7V8Disability
```
Comment (schlick) [inline lib/avetmiss_data/stores/v8/disability.rb]: Extract back out.

File: lib/avetmiss_data/stores/v8/rto_delivery_location.rb [hunk 15]
```diff
@@ -0,0 +1,3 @@
+class AvetmissData::Stores::V8::RtoDeliveryLocation < AvetmissData::Stores::V8::Base
+  include AvetmissData::Stores::V7V8RtoDeliveryLocation
```
Comment (schlick) [inline lib/avetmiss_data/stores/v8/rto_delivery_location.rb]: Extract back out.

File: lib/avetmiss_data/version.rb [hunk 16]
```diff
@@ -1,3 +1,3 @@
 module AvetmissData
-  VERSION = '1.5.0'
+  VERSION = '1.6.0'
```
Comment (schlick) [inline lib/avetmiss_data/version.rb]: With the switch to defaulting to V8, that would be a breaking change so possibly need to go to v2.0.0. See earlier comment.

File: spec/package_spec.rb [hunk 17]
```diff
@@ -32,8 +32,20 @@
...
+        end
+      end
+
+      context 'for a v8 package' do
+        let(:package) { AvetmissData::Package.new(version: :v8) }
+
+        it 'contains all standard files for RTO submission' do
+          expect(file_names).to match_array(AvetmissData::Package::V8_FILES_MAP.values)
+        end
       end
```
Comment (schlick) [inline spec/package_spec.rb]: Spec for V6. Spec unknown/unmatched version.

File: spec/stores/concerns/v7_v8_achievement_spec.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+      context 'and has extra fields' do
+        let(:file_path) { 'spec/fixtures/nat_files/v7/NAT00100_extrafields.txt' }
+        specify { expect(subject.extras).to eq('extra data that we ignore') }
+      end
+    end
+
+    it_behaves_like :store_processing_invalid_file
+  end
+end
```
Comment (schlick) [inline spec/stores/concerns/v7_v8_achievement_spec.rb]: Split

File: spec/stores/concerns/v7_v8_disability_spec.rb [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+      context 'and contains extra fields' do
+        let(:file_path) { 'spec/fixtures/nat_files/v7/NAT00090_extrafields.txt' }
+        specify { expect(subject.extras).to eq('extra data that we ignore') }
+      end
+    end
+
+    it_behaves_like :store_processing_invalid_file
+  end
+end
```
Comment (schlick) [inline spec/stores/concerns/v7_v8_disability_spec.rb]: Split

File: spec/stores/concerns/v7_v8_rto_delivery_location_spec.rb [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+      context 'and contains extra fields' do
+        let(:file_path) { 'spec/fixtures/nat_files/v7/NAT00020_extrafields.txt' }
+        specify { expect(subject.extras).to eq('extra data that we ignore') }
+      end
+    end
+
+    it_behaves_like :store_processing_invalid_file
+  end
+end
```
Comment (schlick) [inline spec/stores/concerns/v7_v8_rto_delivery_location_spec.rb]: Split

File: spec/stores/v7/achievement_spec.rb [hunk 21]
```diff
@@ -1,21 +0,0 @@
...
-      let!(:line) { File.open('spec/fixtures/nat_files/v7/NAT00100.txt').first }
-      subject { AvetmissData::Stores::V7::Achievement.from_line(line) }
-
-      specify { expect(subject).not_to be_blank }
-      specify { expect(subject.client_identifier).to eq('12345') }
-      specify { expect(subject.prior_education_achievement_identifier).to eq('008') }
-      specify { expect(subject.extras).to be_blank }
-    end
-  end
-end
```
Comment (schlick) [inline spec/stores/v7/achievement_spec.rb]: Reinstate.

File: spec/stores/v7/disability_spec.rb [hunk 22]
```diff
@@ -1,21 +0,0 @@
...
-      let!(:line) { File.open('spec/fixtures/nat_files/v7/NAT00090.txt').first }
-      subject { AvetmissData::Stores::V7::Disability.from_line(line) }
-
-      specify { expect(subject).not_to be_blank }
-      specify { expect(subject.client_identifier).to eq('12345') }
-      specify { expect(subject.disability_type_identifier).to eq('13') }
-      specify { expect(subject.extras).to be_blank }
-    end
-  end
-end
```
Comment (schlick) [inline spec/stores/v7/disability_spec.rb]: Reinstate.

File: spec/stores/v7/rto_delivery_location_spec.rb [hunk 23]
```diff
@@ -1,25 +0,0 @@
...
-      specify { expect(subject.training_organisation_delivery_location_identifier).to eq('4707') }
-      specify { expect(subject.training_organisation_delivery_location_name).to eq('Park View Motor Inn-Wangaratta') }
-      specify { expect(subject.post_code).to eq('3677') }
-      specify { expect(subject.state_identifier).to eq('02') }
-      specify { expect(subject.address_location).to eq('WANGARATTA') }
-      specify { expect(subject.country_identifier).to eq('1101') }
-      specify { expect(subject.extras).to be_blank }
-    end
-  end
-end
```
Comment (schlick) [inline spec/stores/v7/rto_delivery_location_spec.rb]: Reinstate.

File: spec/support/store_examples.rb [hunk 24]
```diff
@@ -0,0 +1,10 @@
+shared_examples :store_processing_invalid_file do
+  context 'when the record is invalid' do
+    let!(:file_path) { 'spec/fixtures/nat_files/v8/generic_invalid.txt' }
+
+    specify { expect{ subject }.to_not raise_error }
+    it 'creates a populated store' do
+      expect(subject).not_to be_blank
+    end
+  end
+end
```
Comment (schlick) [inline spec/support/store_examples.rb]: Keep this extraction. It is independent of V7 or V8.

File: lib/avetmiss_data/package.rb (hunk 4 above)
Comment (SyntheticDave) [reply lib/avetmiss_data/package.rb]: V6 and V7 have the same file map. Updated to reflect this.

File: lib/avetmiss_data/package.rb [hunk 25]
```diff
@@ -136,7 +138,8 @@ def self.from_zip_file(zip_file, attributes={})
   end
 
   def to_zip_file
-    files_map = self.version == :v7 ? V7_FILES_MAP : V8_FILES_MAP
+    return unless KNOWN_VERSIONS.include?(self.version)
```
Comment (schlick) [inline lib/avetmiss_data/package.rb]: Perhaps it ought to raise

Comment (SyntheticDave) [reply lib/avetmiss_data/package.rb]: Yeah, could do that. Wasn't sure whether to raise here, or just return a zip with no stores.
I suppose raising actually gives the developer a better idea of what went wrong.

Want me to update this to do so?

**Pull request merged**